### PR TITLE
enable CUDA provider on windows

### PR DESCRIPTION
**Pull request by bratpiorka**
enable CUDA provider on windows

### CODE REVIEW

File: cmake/FindCUDA.cmake [hunk 1]
```diff
@@ -11,7 +11,8 @@ get_filename_component(CUDA_LIB_DIR ${CUDA_LIBRARIES} DIRECTORY)
 set(CUDA_LIBRARY_DIRS ${CUDA_LIB_DIR})
 
 if(WINDOWS)
-    find_file(CUDA_DLL NAMES "bin/cuda.dll" "cuda.dll")
+    find_file(CUDA_DLL NAMES "bin/nvcuda.dll" "nvcuda.dll" "bin/cuda.dll"
+                             "cuda.dll")
```
Comment (vinser52) [inline cmake/FindCUDA.cmake:15]: Do we need booth variants: `bin/nvcuda.dll` and `nvcuda.dll`?

File: test/providers/cuda_helpers.cpp [hunk 2]
```diff
@@ -202,23 +216,40 @@ int cuda_fill(CUcontext context, CUdevice device, void *ptr, size_t size,
...
-int cuda_copy(CUcontext context, CUdevice device, void *dst_ptr, void *src_ptr,
-              size_t size) {
-    (void)context;
+int cuda_copy(CUcontext context, CUdevice device, void *dst_ptr,
+              const void *src_ptr, size_t size) {
     (void)device;
 
+    // set required context
+    CUcontext curr_context = nullptr;
+    set_context(context, &curr_context);
```
Comment (vinser52) [inline test/providers/cuda_helpers.cpp:251]: nit: Do we need to set a context? I though that `cuMemcpy` automatically detects the context of `dst_ptr` and `src_ptr`.

File: test/providers/cuda_helpers.cpp [hunk 3]
```diff
@@ -202,23 +216,40 @@ int cuda_fill(CUcontext context, CUdevice device, void *ptr, size_t size,
         return -1;
     }
 
+    // set required context
+    CUcontext curr_context = nullptr;
+    set_context(context, &curr_context);
```
Comment (vinser52) [inline test/providers/cuda_helpers.cpp:221]: nit: Do we need to set a context? I though that `cuMemsetD32` automatically detects the context of `dst_ptr` and `src_ptr`.

**Review by ldorau — APPROVED** [review]
LGTM

File: test/providers/cuda_helpers.cpp (hunk 2 above)
Comment (bratpiorka) [reply test/providers/cuda_helpers.cpp:251]: test doesn't work without this (only on windows) - cuGetErrorString() returns "invalid argument"

File: test/providers/cuda_helpers.cpp (hunk 3 above)
Comment (bratpiorka) [reply test/providers/cuda_helpers.cpp:221]: test doesn't work without this (only on windows) - cuGetErrorString() returns "invalid argument"

**Review by vinser52 — APPROVED** [review]

File: .github/workflows/reusable_gpu.yml [hunk 4]
```diff
@@ -154,10 +162,53 @@ jobs:
...
+        vcpkgGitCommitId: 3dd44b931481d7a8e9ba412621fa810232b66289
+        vcpkgDirectory: ${{env.BUILD_DIR}}/vcpkg
+        vcpkgJsonGlob: '**/vcpkg.json'
+
+    - name: Install dependencies (windows-latest)
+      if: matrix.os == 'Windows'
+      run: vcpkg install
+      shell: pwsh # Specifies PowerShell as the shell for running the script.
+
+    - name: Install dependencies (ubuntu-latest)
```
Comment (kswiecicki) [inline .github/workflows/reusable_gpu.yml:178]: There's no `ubuntu-latest` in the OS matrix in the `gpu-CUDA` job.

File: test/providers/cuda_helpers.cpp [hunk 5]
```diff
@@ -287,6 +320,25 @@ CUcontext get_current_context() {
     return context;
 }
 
+CUresult set_context(CUcontext required_ctx, CUcontext *restore_ctx) {
```
Comment (kswiecicki) [inline test/providers/cuda_helpers.cpp:323]: This could be `static`, it's not used outside of `cuda_helpers.cpp`.

**Review by lukaszstolarczuk — COMMENTED** [review]
LGTM

File: .github/workflows/reusable_gpu.yml (hunk 4 above)
Comment (lukaszstolarczuk) [reply .github/workflows/reusable_gpu.yml:178]: I'm changing this in my PR (workflows update, so no worries)

Comment (bratpiorka) [reply .github/workflows/reusable_gpu.yml]: removed

File: test/providers/cuda_helpers.cpp (hunk 5 above)
Comment (bratpiorka) [reply test/providers/cuda_helpers.cpp]: done

**Review by kswiecicki — APPROVED** [review]

**Pull request merged**